R/changeLegendColour.R

```diff
@@ -23,38 +23,32 @@ changeLegendColour <- function(x,
   ##Build plot to extract deets from it
   g <- ggplot2::ggplot_build(x)
 
-  if(is.null(g$plot$scales$scales[[1]]$palette.cache)) {
-    stop("You must choose options for the colour/fill in your ggplot2 object.")
-  }
+  ##Check which of the scales contains the colours
+  proto <- c()
+    for(i in 1:length(g$plot$scales$scales)){
 
-  if(is.null(base::names(g$plot$scales$scales[[1]]$palette.cache))) {
-    stop("You must provide names to the vector containing the colours you chose
-         which correspond to the group which is that colour:
-         scale_colour_manual(values = c('group 1' = 'colour 1', ...))")
-  }
+    if(!is.null(g$plot$scales$scales[[i]]$palette.cache)){
 
-  if(is.null(base::names(g$plot$scales$scales[[1]]$palette.cache))) {
-    stop("You must provide names to the vector containing the colours you chose
-         which correspond to the group which is that colour:
-         scale_colour_manual(values = c('group 1' = 'colour 1', ...))")
+      proto <- c(proto, i)
+
+      }
   }
 
-  if(base::length(g$plot$scales$scales[[1]]$labels) > 0 & is.null(base::names(g$plot$scales$scales[[1]]$labels))) {
-    stop("You must provide names to the vector containing the new labels you chose for your legend
-         which correspond to the group which has that new label:
-         scale_colour_manual(values = c('group 1' = 'label 1', ...))")
+  if(is.null(g$plot$scales$scales[[proto]]$palette.cache)) {
+    stop("You must choose options for the colour/fill in your ggplot2 object.")
   }
+
   ##Extract names and colours automagically from plot if available from palette cache
 
   leg_font_colour <- base::data.frame(
-    leg_text = base::names(g$plot$scales$scales[[1]]$palette.cache),
-    col = base::unname(g$plot$scales$scales[[1]]$palette.cache)
+    leg_text = g$plot$scales$scales[[proto]]$get_labels(),
```

Comment:
Issue when you don't have the new legend labels and new colours in the same order is caused through this line. I think it should be leg_text = base::names(g$plot$scales$scales[[proto]]$palette.cache[1:length(g$plot$scales$scales[[proto]]$get_labels())])

Reply:
Yes, example code would be great please!

Reply:
Have resolved this by taking a step back and pulling colours direct from grobs rather than ggplot code; this works regardless of how the colours are set. The two examples we tested + more examples would be useful to add as unit tests